internal/test/deployment/pmm.go

```diff
@@ -0,0 +1,387 @@
+		t:        t,
+
+		pmmVersion:     envs[envVarPMMVersion],
+		pmmServerImage: envs[envVarPMMServerImage],
+		pmmClientImage: envs[envVarPMMClientImage],
+		mongoImage:     envs[envVarMongoImage],
+		mongoTag:       envs[envVarMongoTag],
+
+		useExistingDeployment: useExistingDeployment,
+
+		pmmURL: envs[envVarPMMURL],
+	}
+
+	return pmm
+}
+
+func (p *PMM) PMMURL() string {
+	u, err := url.Parse(p.pmmURL)
+	if err != nil {
+		p.t.Fatal(err)
+	}
+	if u.User.Username() == "" {
+		u.User = url.UserPassword("admin", "admin")
+	}
+	if u.Scheme == "" {
+		u.Scheme = "http"
+	}
+	if strings.Contains(u.Host, ":") {
+		u.Host = u.Host[0:strings.Index(u.Host, ":")]
+	}
+	u.Host += ":" + p.httpPort
+
+	return u.String()
+}
+
+func (p *PMM) UseExistingDeployment() bool {
+	return p.useExistingDeployment
+}
+
+func (p *PMM) ClickhouseURL() string {
+	u, err := url.Parse(p.PMMURL())
+	if err != nil {
+		p.t.Fatal(err)
+	}
+	u.User = nil
+	u.Scheme = "clickhouse"
+	u.Path = "pmm"
+	if strings.Contains(u.Host, ":") {
+		u.Host = u.Host[0:strings.Index(u.Host, ":")]
+	}
+	u.Host += ":" + p.clickhousePort
+
+	return u.String()
+}
+
+func (p *PMM) MongoURL() string {
+	u, err := url.Parse(p.PMMURL())
+	if err != nil {
+		p.t.Fatal(err)
+	}
+	u.Scheme = "mongodb"
+	if strings.Contains(u.Host, ":") {
+		u.Host = u.Host[0:strings.Index(u.Host, ":")]
+	}
+	u.Host += ":" + p.mongoPort
+
+	return u.String()
+}
+
+func (pmm *PMM) Deploy(ctx context.Context) error {
+	if err := pmm.deploy(ctx); err != nil {
+		return errors.Wrap(err, "failed to deploy")
+	}
+	if !pmm.dontCleanup {
+		pmm.t.Cleanup(func() {
+			pmm.Destroy(context.Background())
+		})
+	}
+	return nil
+}
+
+func (pmm *PMM) DontCleanup() {
+	pmm.dontCleanup = true
+}
+
+func (pmm *PMM) SetVersion(version string) {
+	pmm.pmmVersion = version
+}
+
+func (pmm *PMM) Log(args ...any) {
+	pmm.t.Helper()
+
+	args = append([]any{fmt.Sprintf("[pmm] %s:", pmm.testName)}, args...)
+	pmm.t.Log(args...)
+}
+
+var checkImagesMu sync.Mutex
+
+func (pmm *PMM) deploy(ctx context.Context) error {
+	dockerCli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return errors.Wrap(err, "failed to create docker client")
+	}
+	defer dockerCli.Close() //nolint:errcheck
+
+	pmm.Log("Destroying existing deployment")
+	if err := destroy(ctx, filters.NewArgs(filters.Arg("label", PerconaLabel+"="+pmm.testName))); err != nil {
+		return errors.Wrap(err, "failed to destroy existing deployment")
+	}
+
+	pmm.Log("Checking images")
+	checkImagesMu.Lock()
+	for _, image := range []string{pmm.ServerImage(), pmm.ClientImage(), pmm.MongoImage()} {
+		exists, err := ImageExists(ctx, image)
+		if err != nil {
+			return errors.Wrap(err, "failed to check image")
+		}
+		if exists {
+			pmm.Log("Image", image, "exists")
+			continue
+		}
+		pmm.Log("Pulling image", image)
+		if err := PullImage(ctx, image); err != nil {
+			return errors.Wrap(err, "failed to pull image")
+		}
```

Comment:
Shouldn't we release mutes on errors?

Reply:
https://github.com/percona/pmm-dump/pull/99/commits/24e8b9804689c5470b3ca023822a3214eff3d7da